File: 112616_Cans_Бидоны/main.cpp
#include <algorithm>
#include<iostream>
#include<map>
#include<vector>

using namespace std;

int N, K;
vector<int> bidon_volumes;

struct count_of_bidons_and_bidon_last {
    int count_of_bidons;
    int bidon_last;
};

map<int, count_of_bidons_and_bidon_last> count_of_bidons_and_last_bidon_volume_dic;

count_of_bidons_and_bidon_last find_count_for_litres(const int litres) {
    if (count_of_bidons_and_last_bidon_volume_dic.find(litres) != count_of_bidons_and_last_bidon_volume_dic.end())
        return count_of_bidons_and_last_bidon_volume_dic[litres];

    // все уже наполнили
    if (litres == 0) {
        count_of_bidons_and_bidon_last example{};
        example.bidon_last = 0;
        example.count_of_bidons = 0;
        count_of_bidons_and_last_bidon_volume_dic[0] = example;
        return example;
    }
    count_of_bidons_and_bidon_last min_count{};
    min_count.count_of_bidons = 10000;
    min_count.bidon_last = 0;

    for (const int bidon_volume : bidon_volumes)
    {
        if (bidon_volume > litres)
        {
            continue;
        }

        const count_of_bidons_and_bidon_last count_for_litres = find_count_for_litres(litres - bidon_volume);
        if (count_for_litres.count_of_bidons < min_count.count_of_bidons)
        {
            min_count.count_of_bidons = count_for_litres.count_of_bidons;
            min_count.bidon_last = bidon_volume;
        }
    }

    count_of_bidons_and_last_bidon_volume_dic[litres].count_of_bidons = min_count.count_of_bidons + 1;
    count_of_bidons_and_last_bidon_volume_dic[litres].bidon_last = min_count.bidon_last;
    return count_of_bidons_and_last_bidon_volume_dic[litres];
}


int main() {

    cin >> N;
    bidon_volumes.push_back(1);
    bidon_volumes.push_back(5);
    bidon_volumes.push_back(6);
    find_count_for_litres(N);
    const int result_count = count_of_bidons_and_last_bidon_volume_dic[N].count_of_bidons;
    if (result_count == 10001)
    {
        cout << -1;
    }
    else {
        cout << result_count << endl;
        vector<int> result_bidons;
        int n = N;
        for (int i = 0; i < result_count; i++) {
            int bidon = count_of_bidons_and_last_bidon_volume_dic[n].bidon_last;
            result_bidons.push_back(bidon);
            n = n - bidon;
        }

        sort(result_bidons.begin(), result_bidons.end());
        reverse(result_bidons.begin(), result_bidons.end());

        for (const int result_bidon : result_bidons)
        {
            cout << result_bidon << " ";
        }
    }
    return 0;
}

File: 112616_Cans_Бидоны/main.py
import sys

N = int(input())
bidon_volumes = [6, 1, 5]
bidon_volumes = sorted(bidon_volumes)

count_of_bidons_and_last_bidon_volume_dic = {}


def find_count_for_litres(litres):
    if count_of_bidons_and_last_bidon_volume_dic.get(litres) is not None:
        return count_of_bidons_and_last_bidon_volume_dic[litres]

    if litres == 0:  # все уже наполнили
        count_of_bidons_and_last_bidon_volume_dic[0] = 0, 0
        return 0, 0

    min_count = 10000, 0
    for bidon_volume in bidon_volumes:
        if bidon_volume > litres:
            continue

        count_for_litres = find_count_for_litres(litres - bidon_volume)
        if count_for_litres[0] < min_count[0]:
            min_count = count_for_litres[0], bidon_volume
    if min_count[
        0] == 10000:  # не можем подобрать бидоны. например нужно 5 литров собрать только из 3-х литровых бидонов.
        print(0)
        sys.exit()

    count_of_bidons_and_last_bidon_volume_dic[litres] = min_count[0] + 1, min_count[1]
    return count_of_bidons_and_last_bidon_volume_dic[litres]


find_count_for_litres(N)

result_count = count_of_bidons_and_last_bidon_volume_dic[N][0]
print(result_count)

result_bidons = []
n = N
for i in range(result_count):
    bidon = count_of_bidons_and_last_bidon_volume_dic[n][1]
    result_bidons.append(bidon)
    n = n - bidon
result_bidons = sorted(result_bidons, reverse=True)

print(*result_bidons)